nixos/modules/services/networking/ifstate.nix

```diff
@@ -0,0 +1,306 @@
+        description = "The initrd IfState package to use.";
+      };
+
+      settings = lib.mkOption {
+        inherit (settingsFormat) type;
+        default = { };
+        description = "Content of IfState's initrd configuration file. See <https://ifstate.net/2.0/schema/> for details.";
+      };
+
+      cleanupSettings = lib.mkOption {
+        inherit (settingsFormat) type;
+        default = {
+          # required by json schema
+          interfaces = { };
+          # https://codeberg.org/liske/ifstate/issues/118
+          namespaces = { };
+        };
+        description = "Content of IfState's initrd cleanup configuration file. See <https://ifstate.net/2.0/schema/> for details. This configuration gets applied before systemd switches to stage two. The goas is to deconfigurate the whole network in order to prevent access to services, before the firewall is configured. The stage two IfState configuration will start after the firewall is configured.";
+      };
+    };
+  };
+
+  config = lib.mkMerge [
+    (lib.mkIf cfg.enable {
+      assertions = [
+        {
+          assertion = !config.networking.networkmanager.enable;
+          message = "IfState and NetworkManager cannot be used at the same time, as both configure the network in a conflicting manner.";
+        }
+        {
+          assertion = !config.networking.useDHCP;
+          message = "IfState and networking.useDHCP cannot be used at the same time, as both configure the network. Please look into IfState hooks to integrate DHCP: https://codeberg.org/liske/ifstate/issues/111";
+        }
+      ];
+
+      networking.useDHCP = lib.mkDefault false;
+
+      # sane defaults to not let IfState work against the kernel
+      boot.extraModprobeConfig = ''
+        options bonding max_bonds=0
+        options dummy numdummies=0
+        options ifb numifbs=0
+      '';
+
+      environment = {
+        # ifstatecli command should be available to use user, there are other useful subcommands like check or show
+        systemPackages = [ cfg.package ];
+        # match the default value of the --config flag of IfState
+        etc."ifstate/ifstate.yaml".source = settingsFormat.generate "ifstate.yaml" cfg.settings cfg.package;
+      };
+
+      systemd.services.ifstate = {
+        description = "IfState";
+
+        # best effort: many iterations, every change here should be carefully tested
+        # some parts are copied from systemd-networkd
+        wantedBy = [
+          "network.target"
+          "network-online.target"
+          "multi-user.target"
+        ];
+        after = [
+          "network-pre.target"
+          "systemd-sysctl.service"
+          "systemd-modules-load.service"
+        ];
+        before = [
+          "network.target"
+          "network-online.target"
+          "shutdown.target"
+        ];
+        conflicts = [ "shutdown.target" ];
```

Comment:
These should probably be made to match [what `systemd-networkd` does](https://github.com/systemd/systemd/blob/main/units/systemd-networkd.service.in), including its `WantedBy` in its `[Install]` section. See `systemd.special(7)` for explanations of `network-online.target` and `network.target` and how one is "active" and the other is "passive".

`network-online.target` is a little tricky though; `systemd-networkd` implements it by not delaying `network.target` for IPs to be acquired, and having a separate service to delay `network-online.target` until IPs are acquired. How does ifstate handle this sort of thing?

Reply:
`network-online.target` is explicitly included. Because IfState is not a daemon it can not check for connectivity, therefore the target should be reached after IfState is finished starting. There are services depending on `network-online.target` and not `network.target` and therefore these services would fail if IfState would not provide the `network-online.traget`.

Reference to authentik where I've tried to add `network.target` but the maintainer just wanted an exclusive dependency on `network-online.target`: https://github.com/nix-community/authentik-nix/pull/21

Reply:
Will look into that tomorrow.

Reply:
Thanks for the hint, this actually solved problems in two VMs which were not reboot safe till now. I've copied it over.